### PR TITLE
Docs!

### PR DESCRIPTION
**Pull request by jasnell**
Initial docs PR .. still lots of detail to fill in.

### CODE REVIEW

File: doc/api/http2.md [hunk 1]
```diff
@@ -0,0 +1,438 @@
...
+
+Immediately terminates the `Http2Session` and the associated `net.Socket` or
+`tls.TLSSocket`.
+
+#### http2session.destroyed
+
+* Value: {boolean}
+
+Will be `true` if this `Http2Session` instance has been destroyed and must no
+longer be used, otherwise `false.
```
Comment (joyeecheung) [inline doc/api/http2.md]: This is missing a backtick at the end so the following text are not highlighted correctly..

File: doc/api/http2.md [hunk 2]
```diff
@@ -0,0 +1,692 @@
...
+### Class: Http2Session
+
+* Extends: {EventEmitter}
+
+Instances of the `http2.Http2Session` class represent an active communications
+session between an HTTP/2 client and server. Instances of this class are *not*
+intended to be constructed directly by user code.
+
+Every `Http2Session` object is associated with exactly one [`net.Socket`][] or
+[`tls.TLSSocket`][]. Upon creation of the `Http2Session` instance, the `Socket`
```
Comment (cjihrig) [inline doc/api/http2.md]: Could you rephrase the sentence that starts with "Upon". It reads a little awkwardly IMO.

File: doc/api/http2.md [hunk 3]
```diff
@@ -0,0 +1,692 @@
...
+will be associated, with the lifecycle of each being bound to the other. When
+either is destroyed, both will be destroyed.
+
+#### Event: `'close'`
```
Comment (cjihrig) [inline doc/api/http2.md]: Looking at the existing http docs, it doesn't look like backticks are used in the event headers (but the single quotes are).

**Review by cjihrig — COMMENTED** [review]
All of the APIs need the `<!-- YAML` tags.

Did you mean to change the mode on doc/api/index.md?

File: doc/api/http2.md [hunk 4]
```diff
@@ -0,0 +1,692 @@
...
+    ':status': 200
+  });
+  stream.end('<h1>Hello World</h1>');
+});
+
+server.listen(80);
+```
+
+The following illustrates an HTTP/2 client:
+```
```
Comment (kjin) [inline doc/api/http2.md]: The code example here doesn't have `js` highlighting applied to it.

File: doc/api/http2.md [hunk 5]
```diff
@@ -0,0 +1,692 @@
...
+* `callback` {Function}
+
+(TODO: fill in detail)
+
+#### http2session.shutdown(options[, callback])
+
+* `options` {Object}
+  * `graceful` {boolean} `true` to attempt a polite shutdown of the
+    `Http2Session`.
+  * `immediate` (boolean) `true` to force the shutdown to occur immediately,
```
Comment (kjin) [inline doc/api/http2.md]: Nit - should this be {boolean}?

File: doc/api/http2.md [hunk 6]
```diff
@@ -0,0 +1,692 @@
...
+});
+
+let data = '';
+req.setEncoding('utf8');
+req.on('data', (d) => data += d);
+req.on('end', () => client.destroy());
+req.end();
+```
+
+### Class: Http2Session
```
Comment (kjin) [inline doc/api/http2.md:58]: It seems like on the server side one never directly interacts with `Http2Session` objects. Do you think it would be helpful to make a note of that?

File: doc/api/http2.md [hunk 7]
```diff
@@ -0,0 +1,692 @@
...
+The `http2session.type` will be equal to
+`http2.constants.NGHTTP2_SESSION_SERVER` if this `Http2Session` instance is a
+server, and `http2.constants.NGHTTP2_SESSION_CLIENT` if the instance is a
+client.
+
+### Class: Http2Stream
+
+* Extends: {Duplex}
+
+Instances of the `Http2Stream` class represent a bidirectional HTTP/2
```
Comment (kjin) [inline doc/api/http2.md]: Making 'Instances' singular (e.g. 'Each instance') might improve the readability of this sentence. Also, would it be within the scope of this section to give a brief overview of when `Http2Stream` objects are created? For example, to mention that on the client, it's returned by `Http2Session::request`, while on the server, it's passed through the `'stream'` event.

In addition, nit in the last sentence -- 'its lifetime'.

File: doc/api/http2.md [hunk 8]
```diff
@@ -0,0 +1,727 @@
+# HTTP2
+
+> Stability: 1 - Experimental
+
+The `http2` module provides an implementation of the [HTTP/2][] protocol. It
+can be accessed using:
+
+```js
+const http2 = require('http2');
+```
```
Comment (mcollina) [inline doc/api/http2.md:10]: if we are not shipping this in `8.0.0`, we need to ship it behind a flag, as we will be shadowing the `'http2'` module otherwise. Maybe we might do something different here.

File: doc/api/http2.md [hunk 9]
```diff
@@ -0,0 +1,727 @@
...
+
+The `'push'` event is emitted when response headers for a Server Push stream
+are received. The listener callback is passed the [Headers Object][] and flags
+associated with the headers.
+
+```js
+stream.on('push', (headers, flags) => {
+  // ...
+});
+```
```
Comment (mcollina) [inline doc/api/http2.md]: we need an example here.

File: doc/api/http2.md [hunk 10]
```diff
@@ -0,0 +1,727 @@
...
+
+The `'request'` event is emitted when a block of headers associated with an
+HTTP request is received. The listener callback is passed the [Headers Object][]
+and flags associated with the headers.
+
+```js
+stream.on('request', (headers, flags) => {
+  // ...
+});
+```
```
Comment (mcollina) [inline doc/api/http2.md]: and here

File: doc/api/http2.md [hunk 11]
```diff
@@ -0,0 +1,727 @@
...
+```js
+stream.on('response', (headers, flags) => {
+  // ...
+});
+```
+
+This is emitted only when `http2session.type` is equal to
+`http2.constants.NGHTTP_SESSION_CLIENT`.
+
+#### Event: `'streamClosed'
```
Comment (mcollina) [inline doc/api/http2.md]: missing `

File: doc/api/http2.md (hunk 8 above)
Comment (jasnell) [reply doc/api/http2.md:10]: If it's behind a flag (thanks for the reminder btw) then I don't believe that we'll need to do anything different here. We'll just need to document the flag for the time being. I'll work on getting the flag added this week

File: doc/api/http2.md (hunk 9 above)
Comment (jasnell) [reply doc/api/http2.md]: agreed.... but let's add that separately once the push stuff is a bit more settled.

Comment (jasnell) [reply doc/api/http2.md]: In fact, we can keep this PR open for a bit and anyone who would like to push commits against it to fill in examples is more than welcome :-D

**Comment by mcollina** [comment]
I think we should start merging this, and then we can improve it along the way.

**Comment by jasnell** [comment]
@mcollina ... does that count as sign off?

**Review by mcollina — APPROVED** [review]
LGTM

**Comment by jasnell** [comment]
Landed!
